p4-16/spec/P4-16-spec.mdk

```diff
@@ -3471,6 +3485,11 @@ void pop_front(int count) {
 }
 ~ End P4Pseudo
 
+Two header stacks can be compared for equality (==) or inequality (!=)
+only if they have the same element type and the same length.  Two
+stacks are equal if and only if all their corresponding elements are
+equal.
```

Comment:
Actually, one comment here.  Do we want to explicitly say that the 'nextIndex' hidden state that exists at least during parsing time is _not_ relevant for equality comparisons between header stacks?  I know that is implicit in the way it is worded here, but explicit might be better.

Reply:
This makes me realize that the situation is pretty ugly: since assignment is supposed to preserve the nextIndex field, this means that this should field should be even preserved between architectural blocks. So the nextIndex left over in the ingress parser could be accessed in the egress parser! 

It probably should not be used for comparisons, although one could argue either way.

Reply:
In PSA, the only way to preserve an already-parsed header from ingress to egress would be to carry the header in user-defined metadata.  Kind of an odd choice, but I guess not currently disallowed by the architecture spec.

Reply:
Have you considered the idea of allowing constructs that are not in the base P4_16 language spec in the intermediate representations that you can spit out in the compiler?  That would allow you to explicitly define a my_hdr_stack_variable.nextIndex field in that slightly extended language, even if it is not available to P4_16 developers.

Or just go for it and make such a field part of the language explicitly, available everywhere, but not for P4_16 programmer convenience, but because it is helpful in some compilers?